File: android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/auth/handlers/FacebookAuthProviderHandler.java
package dev.robingenz.capacitorjs.plugins.firebase.auth.handlers;

import android.content.Intent;
import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;
import com.facebook.login.widget.LoginButton;
import com.getcapacitor.PluginCall;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FacebookAuthProvider;
import dev.robingenz.capacitorjs.plugins.firebase.auth.FirebaseAuthentication;

public class FacebookAuthProviderHandler {

    public static final int RC_FACEBOOK_AUTH = 105;
    private FirebaseAuthentication pluginImplementation;
    private CallbackManager mCallbackManager;
    private LoginButton loginButton;
    private PluginCall savedCall;

    public FacebookAuthProviderHandler(FirebaseAuthentication pluginImplementation) {
        this.pluginImplementation = pluginImplementation;
        mCallbackManager = CallbackManager.Factory.create();
        loginButton = new LoginButton(pluginImplementation.getPlugin().getContext());
        loginButton.setPermissions("email", "public_profile");
        loginButton.registerCallback(
            mCallbackManager,
            new FacebookCallback<LoginResult>() {
                @Override
                public void onSuccess(LoginResult loginResult) {
                    handleSuccessCallback(loginResult);
                }

                @Override
                public void onCancel() {
                    handleCancelCallback();
                }

                @Override
                public void onError(FacebookException exception) {
                    handleErrorCallback(exception);
                }
            }
        );
    }

    public void signIn(PluginCall call) {
        savedCall = call;
        this.loginButton.performClick();
    }

    public void signOut() {
        LoginManager.getInstance().logOut();
    }

    public void handleOnActivityResult(int requestCode, int resultCode, Intent data) {
        mCallbackManager.onActivityResult(requestCode, resultCode, data);
    }

    private void handleSuccessCallback(LoginResult loginResult) {
        AccessToken token = loginResult.getAccessToken();
        AuthCredential credential = FacebookAuthProvider.getCredential(token.getToken());
        pluginImplementation.handleSuccessfulSignIn(savedCall, credential);
    }

    private void handleCancelCallback() {
        pluginImplementation.handleFailedSignIn(savedCall, "Facebook login canceled.", null);
    }

    private void handleErrorCallback(FacebookException exception) {
        pluginImplementation.handleFailedSignIn(savedCall, null, exception);
    }
}

File: ios/Plugin/Handlers/FacebookAuthProviderHandler.swift
import Foundation
import Capacitor
import FirebaseAuth
#if INCLUDE_FACEBOOK
import FBSDKLoginKit
#endif

class FacebookAuthProviderHandler: NSObject {
    public let errorSignInCanceled = "Sign in canceled."
    private var pluginImplementation: FirebaseAuthentication
#if INCLUDE_FACEBOOK
    private var loginManager: LoginManager
#endif

    init(_ pluginImplementation: FirebaseAuthentication) {
        self.pluginImplementation = pluginImplementation
#if INCLUDE_FACEBOOK
        loginManager = LoginManager()
#endif
        super.init()
    }

    func signIn(call: CAPPluginCall) {
#if INCLUDE_FACEBOOK
        DispatchQueue.main.async {
            self.loginManager.logIn(permissions: ["email", "public_profile"], from: self.pluginImplementation.getPlugin().bridge?.viewController) { result, error in
                if let error = error {
                    self.pluginImplementation.handleFailedSignIn(message: nil, error: error)
                    return
                }

                guard let accessToken = result?.token else {
                    self.pluginImplementation.handleFailedSignIn(message: self.errorSignInCanceled, error: nil)
                    return
                }

                let token = accessToken.tokenString
                let credential = FacebookAuthProvider.credential(withAccessToken: token)
                self.pluginImplementation.handleSuccessfulSignIn(credential: credential, idToken: token, nonce: nil)
            }
        }
#endif
    }

    func signOut() {
#if INCLUDE_FACEBOOK
        loginManager.logOut()
#endif
    }
}
